core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -0,0 +1,157 @@
+ * event log file - caller should keep in mind that this class doesn't care about the semantic of
+ * ordering.
+ *
+ * When compacting the files, the range of compaction for given file list is determined as:
+ * (rightmost compact file ~ the file where there're `maxFilesToRetain` files on the right side)
+ *
+ * If there's no compact file in the list, it starts from the first file. If there're not enough
+ * files after rightmost compact file, compaction will be skipped.
+ */
+class EventLogFileCompactor(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem) extends Logging {
+
+  private val maxFilesToRetain: Int = sparkConf.get(EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN)
+
+  def compact(eventLogFiles: Seq[FileStatus]): Seq[FileStatus] = {
+    if (eventLogFiles.length <= maxFilesToRetain) {
+      return eventLogFiles
+    }
+
+    if (EventLogFileWriter.isCompacted(eventLogFiles.last.getPath)) {
+      return Seq(eventLogFiles.last)
+    }
+
+    val (filesToCompact, filesToRetain) = findFilesToCompact(eventLogFiles)
+    if (filesToCompact.isEmpty) {
+      filesToRetain
+    } else {
+      val builders = EventFilterBuilder.initializeBuilders(fs, filesToCompact.map(_.getPath))
+
+      val rewriter = new FilteredEventLogFileRewriter(sparkConf, hadoopConf, fs,
+        builders.map(_.createFilter()))
+      val compactedPath = rewriter.rewrite(filesToCompact)
+
+      cleanupCompactedFiles(filesToCompact)
+
+      fs.getFileStatus(new Path(compactedPath)) :: filesToRetain.toList
+    }
+  }
+
+  private def cleanupCompactedFiles(files: Seq[FileStatus]): Unit = {
+    files.foreach { file =>
+      var deleted = false
+      try {
+        deleted = fs.delete(file.getPath, true)
+      } catch {
+        case _: IOException =>
+      }
+      if (!deleted) {
+        logWarning(s"Failed to remove ${file.getPath} / skip removing.")
+      }
+    }
+  }
+
+  private def findFilesToCompact(
+      eventLogFiles: Seq[FileStatus]): (Seq[FileStatus], Seq[FileStatus]) = {
+    val files = RollingEventLogFilesFileReader.dropBeforeLastCompactFile(eventLogFiles)
+    if (files.length > maxFilesToRetain) {
+      (files.dropRight(maxFilesToRetain), files.takeRight(maxFilesToRetain))
+    } else {
+      (Seq.empty, files)
+    }
+  }
+}
+
+/**
+ * This class rewrites the event log files into one compact file: the compact file will only
+ * contain the events which pass the filters. Events will be dropped only when all filters
+ * decide to reject the event or don't mind about the event. Otherwise, the original line for
+ * the event is written to the compact file as it is.
+ */
+class FilteredEventLogFileRewriter(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem,
+    filters: Seq[EventFilter]) {
+
+  def rewrite(eventLogFiles: Seq[FileStatus]): String = {
+    require(eventLogFiles.nonEmpty)
+
+    val targetEventLogFilePath = eventLogFiles.last.getPath
```

Comment:
The name of this variable is misleading, since the compacted file will actually have a different name.